### PR TITLE
{Network} `az network vnet`: Migrate to atomic commands

### PR DESCRIPTION
**Pull request by necusjz**
**Related command**

az network vnet

**Description**

Fixes: https://github.com/Azure/azure-cli/issues/24630
Fixes: https://github.com/Azure/azure-cli/issues/23639

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 1]
```diff
@@ -6861,304 +6865,330 @@ def list_traffic_manager_endpoints(cmd, resource_group_name, profile_name, endpo
...
-    ncf = network_client_factory(cmd.cli_ctx, aux_subscriptions=[aux_subscription])
-    return ncf.virtual_network_peerings.begin_create_or_update(
-        resource_group_name, virtual_network_name, virtual_network_peering_name, peering)
+    def post_instance_update(self, instance):
+        if str(instance.properties.nat_gateway.id).split("/")[-1] == "":
+            instance.properties.nat_gateway = None
+        if str(instance.properties.network_security_group.id).split("/")[-1] == "":
+            instance.properties.network_security_group = None
+        if str(instance.properties.route_table.id).split("/")[-1] == "":
+            instance.properties.route_table = None
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:7126]: Recommend using custom `AAZResourceIdArgFormat`

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 2]
```diff
@@ -6861,304 +6865,330 @@ def list_traffic_manager_endpoints(cmd, resource_group_name, profile_name, endpo
...
-    elif service_endpoint_policy:
-        instance.service_endpoint_policies = []
-        for policy in service_endpoint_policy:
-            instance.service_endpoint_policies.append(SubResource(id=policy))
+class VNetUpdate(_VNetUpdate):
+    def post_instance_update(self, instance):
+        if instance.properties.dhcp_options.dns_servers == [""]:
+            instance.properties.dhcp_options.dns_servers = None
+        if str(instance.properties.ddos_protection_plan).split("/")[-1] == "":
+            instance.properties.ddos_protection_plan = None
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:6937]: If we want to deprecate this behavior, recommend handle this in pre_operations with warnings.
Recommend using customize `AAZResourceIdArgFormat`.

File: src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py [hunk 3]
```diff
@@ -5126,15 +5126,15 @@ def test_network_subnet_delegation(self, resource_group):
         self.assertTrue(len(result) > 1, True)
 
         self.cmd('network vnet create -g {rg} -n {vnet} -l westcentralus')
-        self.cmd('network vnet subnet create -g {rg} --vnet-name {vnet} -n {subnet} --address-prefix 10.0.0.0/24 --delegations Microsoft.Web.serverFarms', checks=[
+        self.cmd('network vnet subnet create -g {rg} --vnet-name {vnet} -n {subnet} --address-prefixes 10.0.0.0/24 --delegations Microsoft.Web.serverFarms', checks=[
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py:5129]: why this name changed?

Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/network/tests/latest/test_network_commands.py:5129]: There is no argument named `--address-prefix` https://learn.microsoft.com/en-us/cli/azure/network/vnet/subnet?view=azure-cli-latest#az-network-vnet-subnet-create, i.e., it's intended for Azure Stack.

File: src/azure-cli/azure/cli/command_modules/network/aaz/latest/network/vnet/subnet/_create.py [hunk 4]
```diff
@@ -0,0 +1,2672 @@
...
+            id_part="name",
+        )
+        _args_schema.address_prefix = AAZStrArg(
+            options=["--address-prefix"],
+            help="The address prefix for the subnet.",
+        )
+        _args_schema.address_prefixes = AAZListArg(
+            options=["--address-prefixes"],
+            help="Space-separated list of address prefixes in CIDR format.",
+        )
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/aaz/latest/network/vnet/subnet/_create.py:73]: Can users use --address-prefix togethor with --address-prefixes?

Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/network/aaz/latest/network/vnet/subnet/_create.py:73]: To be honest, users previously can do that. But IMO, it's more like a kind of backdoors as it doesn't meet the document.

File: src/azure-cli/azure/cli/command_modules/network/_help.py [hunk 5]
```diff
@@ -5045,13 +5045,10 @@
...
     populator-commands:
       - az network vnet list-endpoint-services
-  - name: --nat-gateway
-    short-summary: Attach Nat Gateway to subnet
 examples:
   - name: Create new subnet attached to an NSG with a custom route table.
     text: |
-        az network vnet subnet create -g MyResourceGroup --vnet-name MyVnet -n MySubnet \\
-            --address-prefixes 10.0.0.0/24 --network-security-group MyNsg --route-table MyRouteTable
+        az network vnet subnet create -g MyResourceGroup --vnet-name MyVnet -n MySubnet --address-prefixes 10.0.0.0/24 --network-security-group MyNsg --route-table MyRouteTable
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/_help.py:5051]: Do we need to remove the help of vnet related commands?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 6]
```diff
@@ -6861,304 +6866,362 @@ def list_traffic_manager_endpoints(cmd, resource_group_name, profile_name, endpo
...
-        instance.service_endpoint_policies = None
-    elif service_endpoint_policy:
-        instance.service_endpoint_policies = []
-        for policy in service_endpoint_policy:
-            instance.service_endpoint_policies.append(SubResource(id=policy))
+    def post_instance_update(self, instance):
+        if instance.properties.dhcp_options.dns_servers == [""]:
+            instance.properties.dhcp_options.dns_servers = None
+        if instance.properties.ddos_protection_plan == "":
+            instance.properties.ddos_protection_plan = None
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:6957]: You can remove this. It's handled by `EmptyResourceIdArgFormat`. BTW, I think it's ok to ignore the change in generic update as we recommend deprecating "" usage

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 7]
```diff
@@ -6758,304 +6761,390 @@ def list_traffic_manager_endpoints(cmd, resource_group_name, profile_name, endpo
...
+    return Create(cli_ctx=cmd.cli_ctx)(command_args={
+        "name": virtual_network_peering_name,
+        "resource_group": resource_group_name,
+        "vnet_name": virtual_network_name,
+        "remote_vnet": peering["remoteVirtualNetwork"].pop("id", None),
+        "allow_vnet_access": peering.pop("allowVirtualNetworkAccess", None),
+        "allow_gateway_transit": peering.pop("allowGatewayTransit", None),
+        "allow_forwarded_traffic": peering.pop("allowForwardedTraffic", None),
+        "use_remote_gateways": peering.pop("useRemoteGateways", None),
+        "sync_remote": "true",
```
[the remark this answers is not in the document]
Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:7146]: It refers to https://github.com/Azure/azure-rest-api-specs/blob/0b1c0be5c97dcd5dadea4e7f975a556a78f58d60/specification/network/resource-manager/Microsoft.Network/stable/2022-01-01/virtualNetwork.json#L1047-L1060

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 8]
```diff
@@ -6758,304 +6761,390 @@ def list_traffic_manager_endpoints(cmd, resource_group_name, profile_name, endpo
...
-    if disable_private_link_service_network_policies:
-        instance.private_link_service_network_policies = "Disabled"
-    elif disable_private_link_service_network_policies is not None:
-        instance.private_link_service_network_policies = "Enabled"
+    def pre_operations(self):
+        args = self.ctx.args
+        if has_value(args.address_prefixes):
+            prefixes = args.address_prefixes
+            args.address_prefixes = prefixes if len(prefixes) > 1 else None
+            args.address_prefix = prefixes[0] if len(prefixes) == 1 else None
```
[the remark this answers is not in the document]
Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:7038]: prefixes has default value ["10.0.0.0/16"]

**Review by kairu-ms — APPROVED** [review]

**Review by yanzhudd — APPROVED** [review]

**Review by calvinhzy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_create_and_update_with_http_proxy_config.yaml [hunk 9]
```diff
@@ -329,7 +329,7 @@ interactions:
       - AZURECLI/2.42.0 (PIP) azsdk-python-azure-mgmt-network/21.0.1 Python/3.8.9
         (Windows-10-10.0.22621-SP0)
     method: PUT
-    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest000001/providers/Microsoft.Network/virtualNetworks/cliakstest000002?api-version=2022-01-01
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest000001/providers/Microsoft.Network/virtualNetworks/cliakstest000002/subnets/proxy-subnet?api-version=2022-01-01
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_create_and_update_with_http_proxy_config.yaml:332]: I'm curious why `/subnets/proxy-subnet` was added to the `uri`.

**Comment by jsntcy** [comment]
Ensure there is no breaking change in migration.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_create_and_update_with_http_proxy_config.yaml (hunk 9 above)
Comment (necusjz) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_create_and_update_with_http_proxy_config.yaml:332]: Cauz `az network vnet subnet` used VNet API before (kind of subcommand implementation), and it uses its own independent API now.

**Review by jsntcy — APPROVED** [review]

**Pull request merged**